jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/JPAClassFileTransformerProcessor.java

```diff
@@ -38,21 +41,26 @@ private void setClassLoaderTransformer(DeploymentUnit deploymentUnit) {
         // (AS7-2233) each persistence unit can use a persistence provider, that might need
         // to use ClassTransformers.  Providers that need class transformers will add them
         // during the call to CreateContainerEntityManagerFactory.
-
         DelegatingClassTransformer transformer = deploymentUnit.getAttachment(DelegatingClassTransformer.ATTACHMENT_KEY);
-        boolean appContainsPersistenceProviderJars = false;  // remove when we revert WFLY-10520
-        if ( transformer != null) {
-
+        if (transformer != null) {
+            final List<PersistenceUnitMetadata> persistentUnits = new ArrayList<>();
             for (ResourceRoot resourceRoot : DeploymentUtils.allResourceRoots(deploymentUnit)) {
                 PersistenceUnitMetadataHolder holder = resourceRoot.getAttachment(PersistenceUnitMetadataHolder.PERSISTENCE_UNITS);
                 if (holder != null) {
                     for (PersistenceUnitMetadata pu : holder.getPersistenceUnits()) {
-                        if (Configuration.needClassFileTransformer(pu)) {
-                            transformer.addTransformer(new JPADelegatingClassFileTransformer(pu));
+                        if (!Configuration.needClassFileTransformer(pu)) {
+                            // WFLY-20394 Ensure that bytecode enhancement only happens if all persistence units allow bytecode enhancement
```

Comment:
This comment sounds misleading? What if the current PU allows enhancement, but another doesn't?

Reply:
For each application deployment, we are detecting a mix of enhanced + non-enhanced and will disable enhancement for all (EE container managed) persistence units in mixed case.

More concretely, if there are 160 persistence units and 159 have enhancement enabled but 1 doesn't, all 160 persistence units will not be enhanced.  We can revise this logic in the future as needed but IMO I think it matches what we agreed to in https://hibernate.zulipchat.com/#narrow/channel/132094-hibernate-orm-dev/topic/ORM.20bytecode.20enhancement.20.2B.20WildFly

@Sanne does ^ make sense?

Reply:
I'm going to add a warning for the mixed case so users might be informed of it.

Reply:
I made a change to show a warning and correct the check to ensure there is actually a mix of enhanced + not enhanced persistence units.  

Example warning: 
`WARN  [org.jboss.as.jpa] (MSC service thread 1-6) WFLYJPA0076: Some persistence units are configured to be enhanced but persistence unit(s) { 2lcmixedenhancement.jar#second_PU, 2lcmixedenhancement.jar#third_PU } disabled enhancement, none of the persistence units will be enhanced for this application deployment
`

Note that we show the list of persistence units that might need to be changed to allow enhancement so that all of the persistence units will be enhanced.

Reply:
@Sanne I'm going to mark [this](https://github.com/wildfly/wildfly/pull/18757) feedback as resolved.  Please comment again if there are any concerns.

Reply:
> More concretely, if there are 160 persistence units and 159 have enhancement enabled but 1 doesn't, all 160 persistence units will not be enhanced.

Sorry I tried to re-read the thread, but I still don't understand the rationale for disabling enhancement. When it's required, it doesn't seem safe to me to not enhance them. Either we do enhance them all and log a warning about this inconsistency, or you throw an error requiring the inconsistency to be addressed.

Clearly the warning would be less disruptive for existing applications so I expect that is what we'd do here, but perhaps in the future it's worth making it stricter as the side-effects can be rather tricky.

Reply:
Addressed as per discussion today.